### ISSUE TITLE
Reversed Mask Loss

### ISSUE THREAD

**Issue by sergiocasaspastor** [Jireh-Jam/R-MNet-Inpainting-keras#4]
Hi, congratulations on such amazing work.

I have a couple of doubts about the code in the generator_loss function. The paper says that "reversed mask loss compares the squared difference for corresponding pixels specific for regions created by the mask on the image and the reconstructed pixels of the masked-image." 

However, when computing the reversed_mask_loss, the reverse mask (inverse of the mask) is used, comparing only the pixels out of the regions created by the mask. This makes this loss always 0. Therefore, I think maybe we should use directly the mask instead of the inverse of the mask, to compare only the regions inside the mask.

Mask
![mask](https://user-images.githubusercontent.com/39574343/111359121-6780c300-868b-11eb-9617-1b213abcfcae.jpg)

Inverse mask
![00064_test](https://user-images.githubusercontent.com/39574343/111359600-00174300-868c-11eb-8f1c-8327f7a2eb22.png)

If I use this inverse mask, the loss is comparing regions outside of the mask, which always gives 0 for me. If I use the mask, then this reverse mask loss gives a value above 0.

The second doubt is about the weight of each loss in the generator. The paper says that the optimal is for λ=0.4 on reversed mask loss. However, in the code you use 0.4*perceptual_loss + 0.6*reversed_mask_loss, which is the opposite. 

I hope you understand my doubts to reproduce the results.

Thank you very much.

**Comment by meiabdalla** [comment]
I think he means that the masked pixel value will be zero so he used revers mask

**Comment by Jireh-Jam** [comment]
The reverse mask loss compares generated regions only. It does not compare the regions outside the mask. The pixel regions you want to keep is 1 and the mask regions is 0. When you reverse the mask on the image, the mask pixel regions become 1 and the rest of the image pixel regions become 0. 
However, I have made changes to the generator loss. 
If you are trying to reproduce the code your own way, you have to make sure the reversed mask when applied on the image shows only the valid regions. Be sure to view the image before proceeding to compile and run the code. Otherwise follow the instructions. I will upload some pertained weights by the first week of January 2022.

**Comment by m-elhussieny** [comment]
for the generator you use mask pixels as 1 so, you multiply image with reverse mask (use the whole image as input and masked pixels will be zero). but in the loss you multiply also the reverse mask means that you compare the are out of mask

**Comment by Jireh-Jam** [comment]
def generator_loss(self, y_true,y_pred):
        ones = K.ones_like(y_pred[:,:,:,1])
        mask = K.stack([ones]*self.channels, axis=-1)
        input_img = Lambda(lambda x : x[:,:,:,0:3])(y_true)
        output_img = Lambda(lambda x : x[:,:,:,0:3])(y_pred)
        reversed_mask = Lambda(self.reverse_mask,output_shape=(self.img_shape_mask))(mask)
        vgg = VGG19(include_top=False, weights='imagenet', input_shape=self.img_shape)
        loss_model = Model(inputs=vgg.input, outputs=vgg.get_layer('block3_conv3').output)
        loss_model.trainable = False
        p_loss = K.mean(K.square(loss_model(output_img) - loss_model(input_img)))
        masking = Multiply()([reversed_mask,input_img])
        predicting = Multiply()([reversed_mask, output_img])
        reversed_mask_loss = (K.mean(K.square(loss_model(predicting) - loss_model(masking))))
        new_loss = 0.6*(p_loss) + 0.4*reversed_mask_loss
        return new_loss

**Comment by Jireh-Jam** [comment]
Pre-trained model is ready. You can use this for inpainting.

**Comment by meiabdalla** [comment]
Thanks

**Comment by white43** [comment]
@Jireh-Jam Apologize for necroposting. I'm confused. Your words in this thread say '_The reverse mask loss compares generated regions only_,' which means 1) the function compares the same regions and 2) the function compares only target (inpainting) regions. Next, in your article you write '_Our reversed mask loss compares the squared difference for corresponding pixels specific for regions created by the mask on the image and the reconstructed pixels of the masked-image_,' (right above Formula 4), and these words appear consistent with the previous ones.

However, in Formula 4 you compare different regions: $I_{M pred}$ and $I \odot  M_r$. In the left part of the formula, you use the original mask $M$. While in the right part you take $M_r$, which is a reversed mask. In other words, the formula compares target (inpainting) regions in restored images with source regions in original images.

Next, in your code, you compare the same regions. This is consistent with your words but not with the formula.

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L144-L146

However, despite being consistent with your words, your current code appears incorrect since it compares source regions rather than target (inpainting) regions. Let me explain why. My masks are white in target area and black in source area, so I have removed the subtraction from 1 here:

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L280-L282

Then, during debugging, I found out that `mask` variable already contains my original mask here, which is white in target area (target area is `1`, source area is `0`):

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L133

In the next line you create `reversed_mask` (target area is `0`, source area is `1`):

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L134

And finally, `reversed_mask` is used to get the regions for further comparison. But since `reversed_mask` has values `1` in source area and `0` in target area, you compare two source regions here:

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L144-L146

When in the code below I remove the left part and the coefficient `0.4` on the right, I get loss values about 0.5-1 during the first epoch with `reversed_mask`. But if I use `mask` instead of `reversed_mask`, my loss values jump to 25-35. For the first epoch and with no pretrained models, the second variant appears more realistic.

https://github.com/Jireh-Jam/R-MNet-Inpainting-keras/blob/864546d68aa3dc070c0b57e923ab1074e330492a/rmnet.py#L147

Since I'm working on a project that uses your code, I will be happy to open a pull request if I understand the issue correctly. It seems we don't need to use `reversed_mask` in `generator_loss`.

Based on the above, I have a few questions:

1. Do I really need to remove the subtraction from 1 when my target area is white and source area is black (like the first image in the first post)?
2. Do we need to compare the same regions (as you say) or different regions (as in Formula 4 in your article)?

Thank you.
